src/js/common.ml

```diff
@@ -157,25 +157,116 @@ let js_parse_motoko s =
     end)
     in Js.some (js_of_sexpr (Arrange.prog prog)))
 
-let js_parse_motoko_typed paths =
-  let paths = paths |> Js.to_array |> Array.to_list in
-  let load_result = Mo_types.Cons.session (fun _ ->
-    Pipeline.load_progs Pipeline.parse_file (paths |> List.map Js.to_string) Pipeline.initial_stat_env)
+let js_parse_motoko_with_deps path s =
+  let main_file = Js.to_string path in
+  let s = Js.to_string s in
+  let prog_and_deps_result =
+    let open Diag.Syntax in
+    let* prog, _ = Pipeline.parse_string main_file s in
+    let* deps =
+      Pipeline.ResolveImport.resolve (Pipeline.resolve_flags ()) prog main_file
+    in
+    Diag.return (prog, deps)
   in
-  js_result load_result (fun (libs, progs, senv) ->
-  progs |> List.map (fun prog ->
+  js_result prog_and_deps_result (fun (prog, deps) ->
     let open Mo_def in
-    let module Arrange_sources_types = Arrange.Make (struct
+    let module Arrange = Arrange.Make (struct
       let include_sources = true
-      let include_types = true
+      let include_types = false
       let include_docs = Some prog.note.Syntax.trivia
       let include_parenthetical = false
-      let main_file = Some prog.at.left.file
-    end)
-    in object%js
-      val ast = js_of_sexpr (Arrange_sources_types.prog prog)
-      (* val typ = js_of_sexpr (Arrange_sources_types.typ typ) *)
-    end) |> Array.of_list |> Js.array |> Js.some)
+      let main_file = Some main_file
+    end) in
+    Js.some (
+      object%js
+        val ast = js_of_sexpr (Arrange.prog prog)
+        val immediateImports =
+          deps
+          |> List.map (fun dep -> Js.string (Pipeline.resolved_import_name dep))
+          |> Array.of_list
+          |> Js.array
+      end))
+
+module Map_conversion (Map : Map.S) = struct
+  let from_js
+    (type data)
+    (map : _ Js.t)
+    (from_key : 'k Js.t -> Map.key)
+    (from_data : 'd Js.t -> data)
+    : data Map.t
+  =
+  let result = ref Map.empty in
+  let callback =
+    (* [forEach] in JS gives value, key, and map, in this order. *)
+    Js.wrap_callback (fun v k _m ->
+      let k = from_key k in
+      let v = from_data v in
+      result := Map.add k v !result)
+  in
+  ignore (Js.Unsafe.meth_call map "forEach" [|Js.Unsafe.inject callback|]);
+  !result
+
+  let to_js
+    (type data)
+    (map : data Map.t)
+    (from_key : Map.key -> Js.Unsafe.top Js.t)
+    (from_data : data -> Js.Unsafe.top Js.t)
+    : _ Js.t
+  =
+  let js_map = Js.Unsafe.new_obj (Js.Unsafe.pure_js_expr "Map") [||] in
+  Map.iter
+    (fun k v ->
+      ignore (Js.Unsafe.meth_call js_map "set" [| from_key k; from_data v |]))
+    map;
+  js_map
+
+  let from_ocaml = to_js
+  let to_ocaml = from_js
+end
+
+let js_parse_motoko_typed paths scope_cache =
```

Comment:
Perhaps `scope_cache` could be optional for backwards-compatibility? We could do this on the `node-motoko` side by passing an empty map, but it would be ideal to preserve the original behavior when directly calling the `moc.js` function.

Reply:
Sure. Could you clarify the rules/intentions regarding backward compatibility? So, having `scope_cache` optional allows us to run the older VSC extension with a newer compiler version.

Should we check some other cases?

Reply:
What I'm thinking is that when passing only one argument to `parseMotokoTyped()` in `node-motoko`, we should preserve the original behavior (including return type) to minimize the chance of breaking downstream applications. This may not be possible with `moc.js`, so we could perhaps accept the breaking change here and then maintain the original behavior on the `node-motoko` side. 

I also opened https://github.com/serokell/motoko/pull/44 as a starting point for testing the `parseMotokoTyped()` behavior in `moc.js`. Feel free to expand on this, although the tests in `node-motoko` are a higher priority imo.

Reply:
Would the other option be to just add some new entry point and keep the old ones?

Reply:
If we eventually wanted to support switching between different versions in the extension, it might be simpler to check whether the method uses the new function interface if it used a different name. Either way should be ok, although we would err on the side of using a new function for this.

Reply:
So, should we do this? I mean make a new API function instead ?

Reply:
Yes, that would be best as long as this doesn't introduce too much extra complexity to the codebase.

Reply:
Done, you can check it out